### PR TITLE
Changes in `injbo_optimize_rcds` made during machine study

### PR DESCRIPTION
**Pull request by vellosok75**
Changes from 2024-05-27 Machine Study (Optimzation of InjBO)
- add klystron2 amplitude & phases and rf amplitude & phase during bottom ramp as optimization knobs;
- fix some bugs; 

OBS:
- @xresende had suggested some changes in the logic for setting knobs and limits. I initially accepted them, but now decided to keep the old format
- this PR initially also addressed some issues regarding an undefined variable in RCDS. This issue is now addressed in #275

### CODE REVIEW

**Review by fernandohds564 — APPROVED** [review]

File: apsuite/commisslib/injbo_optimize_rcds.py [hunk 1]
```diff
@@ -67,7 +119,7 @@ def __init__(self, isonline=True, use_thread=True):
         """."""
         _RCDS.__init__(self, isonline=isonline, use_thread=use_thread)
         self.params = OptimizeInjBOParams()
-        self.data['strengths'] = []
```
Comment (xresende) [inline apsuite/commisslib/injbo_optimize_rcds.py:70]: should you not use a more general expression, `knob_value`, `param_value` ?

File: apsuite/commisslib/injbo_optimize_rcds.py [hunk 2]
```diff
@@ -16,10 +16,53 @@ class OptimizeInjBOParams(_RCDSParams):
     """."""
 
     KNOBS = [
-        'li_qf3', 'tb_qf2a', 'tb_qf2b', 'tb_qd2a', 'tb_qd2b', 'posx', 'angx',
-        'posy', 'angy', 'kckr']
-    LIMS_UPPER = [+3.0, 9.5, 6.0, 8.5, 5.5, +2.0, +1.0, +2.0, +1.0, -25.0]
-    LIMS_LOWER = [-3.0, 5.0, 2.0, 4.0, 1.5, -2.0, -1.0, -2.0, -1.0, -19.0]
+        'li_qf3',
```
Comment (xresende) [inline apsuite/commisslib/injbo_optimize_rcds.py:19]: @vellosok75 , I find this grouping of KNOB parameters unnatural and more difficult to read. I created a branch `OptimizeInjBOParams` with suggestions. please take a look at it and if you agree, merge it into this PR, if not delete the branch.

**Review by xresende — APPROVED** [review]

**Review by otavio-silveira — APPROVED** [review]

**Review by murilobalves — APPROVED** [review]

File: apsuite/commisslib/injbo_optimize_rcds.py [hunk 3]
```diff
@@ -135,26 +197,34 @@ def get_current_position(self):
...
+                pos.append(self.devices['li_llrf'].dev_klystron2.amplitude)
+            elif knob.lower().startswith('kly2_phs'):
+                pos.append(self.devices['li_llrf'].dev_klystron2.phase)
+            elif knob.lower().startswith('borf_amp'):
+                pos.append(self.devices['bo_llrf'].voltage_bottom)
+            elif knob.lower().startswith('borf_phs'):
+                pos.append(self.devices['bo_llrf'].phase_bottom)
             else:
                 raise ValueError('Wrong specification of knob.')
         return _np.array(pos)
```
Comment (fernandohds564) [inline apsuite/commisslib/injbo_optimize_rcds.py:230]: maybe we could reduce the repetition here this way:
```suggestion
        for knob in self.params.knobs:
            fun = knob.lower().startswith
            if fun('li_qf3'):
                pos.append(self.devices['li_qf3'].current)
            elif fun('tb_qf2a'):
                pos.append(self.devices['tb_qf2a'].current)
            elif fun('tb_qf2b'):
                pos.append(self.devices['tb_qf2b'].current)
            elif fun('tb_qd2a'):
                pos.append(self.devices['tb_qd2a'].current)
            elif fun('tb_qd2b'):
                pos.append(self.devices['tb_qd2b'].current)
            elif fun('posx'):
                pos.append(self.devices['pos_ang'].delta_posx)
            elif fun('angx'):
                pos.append(self.devices['pos_ang'].delta_angx)
            elif fun('posy'):
                pos.append(self.devices['pos_ang'].delta_posy)
            elif fun('angy'):
                pos.append(self.devices['pos_ang'].delta_angy)
            elif fun('kckr'):
                pos.append(self.devices['injkckr'].strength)
            elif fun('kly2_amp'):
                pos.append(self.devices['li_llrf'].dev_klystron2.amplitude)
            elif fun('kly2_phs'):
                pos.append(self.devices['li_llrf'].dev_klystron2.phase)
            elif fun('borf_amp'):
                pos.append(self.devices['bo_llrf'].voltage_bottom)
            elif fun('borf_phs'):
                pos.append(self.devices['bo_llrf'].phase_bottom)
            else:
                raise ValueError('Wrong specification of knob.')
        return _np.array(pos)
```

File: apsuite/commisslib/injbo_optimize_rcds.py [hunk 4]
```diff
@@ -166,26 +236,34 @@ def set_position_to_machine(self, pos):
...
+            elif knob.lower().startswith('kly2_amp'):
+                self.devices['li_llrf'].dev_klystron2.amplitude = p
+            elif knob.lower().startswith('kly2_phs'):
+                self.devices['li_llrf'].dev_klystron2.phase = p
+            elif knob.lower().startswith('borf_amp'):
+                self.devices['bo_llrf'].voltage_bottom = p
+            elif knob.lower().startswith('borf_phs'):
+                self.devices['bo_llrf'].phase_bottom = p
             else:
                 raise ValueError('Wrong specification of knob.')
```
Comment (fernandohds564) [inline apsuite/commisslib/injbo_optimize_rcds.py:268]: same here
```suggestion
        for p, knob in zip(pos, self.params.knobs):
            fun = knob.lower().startswith
            if fun('li_qf3'):
                self.devices['li_qf3'].current = p
            elif fun('tb_qf2a'):
                self.devices['tb_qf2a'].current = p
            elif fun('tb_qf2b'):
                self.devices['tb_qf2b'].current = p
            elif fun('tb_qd2a'):
                self.devices['tb_qd2a'].current = p
            elif fun('tb_qd2b'):
                self.devices['tb_qd2b'].current = p
            elif fun('posx'):
                self.devices['pos_ang'].delta_posx = p
            elif fun('angx'):
                self.devices['pos_ang'].delta_angx = p
            elif fun('posy'):
                self.devices['pos_ang'].delta_posy = p
            elif fun('angy'):
                self.devices['pos_ang'].delta_angy = p
            elif fun('kckr'):
                self.devices['injkckr'].strength = p
            elif fun('kly2_amp'):
                self.devices['li_llrf'].dev_klystron2.amplitude = p
            elif fun('kly2_phs'):
                self.devices['li_llrf'].dev_klystron2.phase = p
            elif fun('borf_amp'):
                self.devices['bo_llrf'].voltage_bottom = p
            elif fun('borf_phs'):
                self.devices['bo_llrf'].phase_bottom = p
            else:
                raise ValueError('Wrong specification of knob.')
```

**Review by fernandohds564 — COMMENTED** [review]
@vellosok75, I've made a few suggestions in the code.

**Comment by vellosok75** [comment]
> @vellosok75, I've made a few suggestions in the code.

Thank you, Fernando. I tried to apply the suggestions directly from here, but apparently Github has problems with doing that when there is line deletion (at least it's what the error message said). Anyways, I have commited your suggestions manually.

**Pull request merged**